File: backend/src/main/java/io/turismo/backend/model/TouristSpot.java
package io.turismo.backend.model;

import jakarta.persistence.*;
import lombok.*;
import org.locationtech.jts.geom.Point;
import java.time.LocalTime;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "tourist_spots")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class TouristSpot {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "tourist_spot_id")
    private UUID touristSpotId;

    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "geometry(Point, 4326)")
    private Point location;

    @Column(name = "opens_at")
    private LocalTime opensAt;

    @Column(name = "closes_at")
    private LocalTime closesAt;

    @Column(length = 255)
    private String shortDescription;

    @Column(columnDefinition = "TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "spot_manager_id", nullable = false)
    private SpotManager spotManager;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "city_id", nullable = false)
    private City city;

    @OneToMany(mappedBy = "touristSpot", cascade = CascadeType.ALL, fetch = FetchType.LAZY, orphanRemoval = true)
    private List<Photo> photos;

    @Builder.Default
    @OneToMany(mappedBy = "touristSpot", cascade = CascadeType.ALL, fetch = FetchType.LAZY, orphanRemoval = true)
    private Set<Activity> activities = new HashSet<>();

    @OneToMany(mappedBy = "touristSpot", cascade = CascadeType.ALL, fetch = FetchType.LAZY, orphanRemoval = true)
    private List<Warn> warns;

    @ManyToMany
    @JoinTable(
        name = "tourist_spot_tags",
        joinColumns = @JoinColumn(name = "tourist_spot_id"),
        inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    private List<Tag> tags;

    @ManyToMany
    @JoinTable(
        name = "tourist_spot_tour_guides",
        joinColumns = @JoinColumn(name = "tourist_spot_id"),
        inverseJoinColumns = @JoinColumn(name = "tour_guide_id")
    )
    private List<TourGuide> tourGuides;
}

File: frontend/src/app/core/models/tourist-spot.ts
import { City } from './city';
import { Photo } from './photo';
import { SpotManagerSimple } from './spot-manager';
import { Tag } from './tag';
import { TourGuideSimple } from './tour-guide';
import { Warn } from './warn';
import { SocialMedia, SocialMediaCreate } from './social-media';

export interface Point {
  type: string;
  coordinates: number[];
}

export interface TouristSpotCreate {
  name: string;
  latitude: number;
  longitude: number;
  opensAt: string;
  closesAt: string;
  shortDescription: string;
  description: string;
  cityId: string;
  tags?: string[];
  socialsMedia?: SocialMediaCreate[];
}

export interface TouristSpot {
  touristSpotId: string;
  name: string;
  location: Point;
  opensAt: string;
  closesAt: string;
  shortDescription: string;
  description: string;
  spotManager: SpotManagerSimple;
  city: City;
  photos: Photo[];
  warns: Warn[];
  tags: Tag[];
  tourGuides: TourGuideSimple[];
  socialsMedia: SocialMedia[];
}

export interface TouristSpotList {
  touristSpotId: string;
  name: string;
  opensAt: string;
  closesAt: string;
  shortDescription: string;
  city: City;
  tags: Tag[];
  photos: Photo[];
  distance?: number;
}

export interface TouristSpotOnMap {
  touristSpotId: string;
  name: string;
  location: Point;
}

export interface TouristSpotToMap {
  touristSpotId: string;
  name: string;
}

export interface TouristSpotUpdate {
  name?: string;
  latitude?: number;
  longitude?: number;
  opensAt?: string;
  closesAt?: string;
  shortDescription?: string;
  description?: string;
  cityName?: string;
  tags?: string[];
  socialsMedia?: SocialMediaCreate[];
}

export interface TouristSpotFilters {
  name?: string | null;
  cityName?: string | null;
  stateName?: string | null;
  tags?: string[] | null;
  longitude?: number | null;
  latitude?: number | null;
  radius?: number | null;
  distance?: number | null;
}
